pkg/pillar/cmd/upgradeconverter/movepatchenvelopes.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package upgradeconverter
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	fileutils "github.com/lf-edge/eve/pkg/pillar/utils/file"
+)
+
+const (
+	srcDir  = "/zedrouter/AppInstMetaData"
+	destDir = "/msrv/AppInstMetaData"
+)
+
+func exists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+func movePersistPubsub(ctxPtr *ucContext) error {
+	src := filepath.Join(ctxPtr.persistStatusDir, srcDir)
+	dst := filepath.Join(ctxPtr.persistStatusDir, destDir)
+
+	srcExists, err := exists(src)
+	if err != nil {
+		return fmt.Errorf("stat zedrouter failed %v", err)
+	}
+
+	dstExists, err := exists(dst)
+	if err != nil {
+		return fmt.Errorf("stat msrv failed %v", err)
+	}
+
+	if srcExists {
+		if dstExists {
+			// if there is already a folder we are not going to
+			// override it
```

Comment:
Typically msrv and the other microservices will start after upgradeconverter is done. But I don't know if we should assume that the existence of an empty directory means that the upgrade has already been done just because /persist/status/msrv/AppInstanceMetadata exists and is empty.

I haven't checked, but it could be that other upgrade path looks at each file and avoids clobbering if the file exists.
That is necessary should there be an upgrade followed by a fallback/downgrade, and then an upgrade again. So please check the other upgrade code.

Reply:
Yes, I didn't think about fallback, so I'm leaving zedrouter pubsub there :)